src/buffer/out/textBuffer.cpp

```diff
@@ -2888,125 +2900,400 @@ std::vector<til::point_span> TextBuffer::SearchText(const std::wstring_view& nee
+std::vector<std::wstring> TextBuffer::Commands() const
 {
-    if (_marks.size() == 0)
+    std::vector<std::wstring> commands{};
+    const auto bottom = _estimateOffsetOfLastCommittedRow();
+    auto lastPromptY = bottom;
+    for (auto promptY = bottom; promptY >= 0; promptY--)
     {
-        return L"";
+        const auto& currRow = GetRowByOffset(promptY);
+        auto& rowPromptData = currRow.GetScrollbarData();
+        if (!rowPromptData.has_value())
+        {
+            // This row didn't start a prompt, don't even look here.
+            continue;
+        }
+
+        // This row did start a prompt! Find the prompt that starts here.
+        // Presumably, no rows below us will have prompts, so pass in the last
+        // row with text as the bottom
+        auto foundCommand = _commandForRow(promptY, lastPromptY);
+        if (!foundCommand.empty())
+        {
+            commands.emplace_back(std::move(foundCommand));
+        }
+        lastPromptY = promptY;
     }
+    std::reverse(commands.begin(), commands.end());
+    return commands;
+}
 
-    const auto& curr{ _marks.back() };
-    const auto& start{ curr.end };
-    const auto& end{ GetCursor().GetPosition() };
+void TextBuffer::StartPrompt()
+{
+    const auto currentRowOffset = GetCursor().GetPosition().y;
+    auto& currentRow = GetMutableRowByOffset(currentRowOffset);
+    currentRow.StartPrompt();
 
-    const auto line = start.y;
-    const auto& row = GetRowByOffset(line);
-    return row.GetText(start.x, end.x);
+    auto attr = GetCurrentAttributes();
+    attr.SetMarkAttributes(MarkKind::Prompt);
+    SetCurrentAttributes(attr);
 }
 
-void TextBuffer::SetCurrentPromptEnd(const til::point pos) noexcept
+bool TextBuffer::_createPromptMarkIfNeeded()
 {
-    if (_marks.empty())
+    // We might get here out-of-order, without seeing a StartPrompt (FTCS A)
+    // first. Since StartPrompt actually sets up the prompt mark on the ROW, we
+    // need to do a bit of extra work here to start a new mark (if the last one
+    // wasn't in an appropriate state).
+
+    const auto mostRecentMarks = GetMarkExtents(1u);
+    if (!mostRecentMarks.empty())
     {
-        return;
+        const auto& mostRecentMark = til::at(mostRecentMarks, 0);
+        if (mostRecentMark.HasOutput())
+        {
+            // The most recent command mark had output. That suggests that either:
+            // * shell integration wasn't enabled (but the user would still
+            //   like lines with enters to be marked as prompts)
+            // * or we're in the middle of a command that's ongoing.
+
+            // If the mark doesn't have any command - then we know we're
+            // playing silly games with just marking whole lines as prompts,
+            // then immediately going to output.
+            //   --> add a new mark to this row, set all the attrs in this
+            //   row to be Prompt, and set the current attrs to Output.

```

Comment:
fwiw the _caller_ does this part, right? I didn't see attribute manipulation code in this function

Reply:
Yep, clarified that.